groth16/src/ext_wit.rs

```diff
@@ -1,12 +1,117 @@
 use ark_ff::{FftField, PrimeField};
+use ark_poly::EvaluationDomain;
+use ark_std::cfg_into_iter;
 use dist_primitives::channel::MpcSerNet;
 use dist_primitives::dfft::{d_fft, d_ifft};
+use dist_primitives::utils::pack::{pack_vec, transpose};
 use mpc_net::{MpcNetError, MultiplexedStreamID};
 use rand::Rng;
 use secret_sharing::pss::PackedSharingParams;
 
+use crate::qap::PackedQAPShare;
 use crate::ConstraintDomain;
 
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+pub async fn h<
+    F: FftField + PrimeField,
+    D: EvaluationDomain<F>,
+    Net: MpcSerNet,
+>(
+    qap_share: PackedQAPShare<F, D>,
+    pp: &PackedSharingParams<F>,
+    cd: &ConstraintDomain<F>,
+    net: &Net,
+) -> Result<Vec<F>, MpcNetError> {
+    const CHANNEL0: MultiplexedStreamID = MultiplexedStreamID::Zero;
+
+    let p_coeff = d_ifft(
+        qap_share.a,
+        true,
+        2,
+        false,
+        &cd.constraint,
+        pp,
+        net,
+        CHANNEL0,
+    )
+    .await?;
+
+    let p_eval =
+        d_fft(p_coeff, false, 1, false, &cd.constraint2, pp, net, CHANNEL0)
+            .await?;
+    let q_coeff = d_ifft(
+        qap_share.b,
+        true,
+        2,
+        false,
+        &cd.constraint,
+        pp,
+        net,
+        CHANNEL0,
+    )
+    .await?;
+
+    let q_eval =
+        d_fft(q_coeff, false, 1, false, &cd.constraint2, pp, net, CHANNEL0)
+            .await?;
+
+    let w_coeff = d_ifft(
+        qap_share.c,
+        true,
+        2,
+        false,
+        &cd.constraint,
+        pp,
+        net,
+        CHANNEL0,
+    )
+    .await?;
+
+    let w_eval =
+        d_fft(w_coeff, false, 1, false, &cd.constraint2, pp, net, CHANNEL0)
+            .await?;
```

Comment:
Should be able to parallelize these operations with `tokio::try_join!`. See examples in the repo.

Reply:
done